Compiler error C4703

Sorry, but with the last modification the compiler issues the error C4703.
To avoid the error, I have put all the code following the "if (hKernel32) ..." but the return(0) line, inner to the "if".

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -5,3 +5,4 @@ project(sys_time_h LANGUAGES C)
 add_executable(example example.c sys/time.h)
 
 target_include_directories(example PRIVATE ${CMAKE_SOURCE_DIR})
+target_compile_options(example PRIVATE "/W4 /WX")

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -5,6 +5,8 @@ project(sys_time_h LANGUAGES C)
 add_executable(example example.c sys/time.h)
 
 target_include_directories(example PRIVATE ${CMAKE_SOURCE_DIR})
-target_compile_options(example PRIVATE
-  $<$<CXX_COMPILER_ID:MSVC>:/W4 /WX>
-  $<$<NOT:$<CXX_COMPILER_ID:MSVC>>:-Wall -Wextra -Wpedantic -Werror>)
+if(MSVC)
+  target_compile_options(example PRIVATE /W4 /WX)
+else(MSVC)
+  target_compile_options(example PRIVATE -Wall -Wextra -Wpedantic -Werror)
+endif(MSVC)

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -5,4 +5,6 @@ project(sys_time_h LANGUAGES C)
 add_executable(example example.c sys/time.h)
 
 target_include_directories(example PRIVATE ${CMAKE_SOURCE_DIR})
-target_compile_options(example PRIVATE "/W4 /WX")
+target_compile_options(example PRIVATE
+  $<$<CXX_COMPILER_ID:MSVC>:/W4 /WX>
+  $<$<NOT:$<CXX_COMPILER_ID:MSVC>>:-Wall -Wextra -Wpedantic -Werror>)

--- a/sys/time.h
+++ b/sys/time.h
@@ -52,8 +52,15 @@ static int gettimeofday(struct timeval *tp, struct timezone *tzp)
 	pfnGetSystemTimePreciseAsFileTime fnGetSystemTimePreciseAsFileTime = NULL;
 	FILETIME time;
 	hKernel32 = GetModuleHandleW(L"kernel32.dll");
+#if defined(__GNUC__)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wcast-function-type"
+#endif
 	if (hKernel32)
 		fnGetSystemTimePreciseAsFileTime = (pfnGetSystemTimePreciseAsFileTime) GetProcAddress(hKernel32, "GetSystemTimePreciseAsFileTime");
+#if defined(__GNUC__)
+#pragma GCC diagnostic pop
+#endif
 
 	if (fnGetSystemTimePreciseAsFileTime)
 		fnGetSystemTimePreciseAsFileTime(&time);

--- a/sys/time.h
+++ b/sys/time.h
@@ -48,8 +48,8 @@ struct timezone
 static int gettimeofday(struct timeval *tp, struct timezone *tzp)
 {
 	typedef void (__stdcall * pfnGetSystemTimePreciseAsFileTime)(LPFILETIME lpSystemTimeAsFileTime);
-	HMODULE hKernel32;
-	pfnGetSystemTimePreciseAsFileTime fnGetSystemTimePreciseAsFileTime;
+	HMODULE hKernel32 = NULL;
+	pfnGetSystemTimePreciseAsFileTime fnGetSystemTimePreciseAsFileTime = NULL;
 	FILETIME time;
 	hKernel32 = GetModuleHandleW(L"kernel32.dll");
 	if (hKernel32)

--- a/sys/time.h
+++ b/sys/time.h
@@ -52,13 +52,13 @@ static int gettimeofday(struct timeval *tp, struct timezone *tzp)
 	pfnGetSystemTimePreciseAsFileTime fnGetSystemTimePreciseAsFileTime = NULL;
 	FILETIME time;
 	hKernel32 = GetModuleHandleW(L"kernel32.dll");
-#if defined(__GNUC__)
+#if defined(__GNUC__) && (__GNUC__ == 8)
 #pragma GCC diagnostic push
 #pragma GCC diagnostic ignored "-Wcast-function-type"
 #endif
 	if (hKernel32)
 		fnGetSystemTimePreciseAsFileTime = (pfnGetSystemTimePreciseAsFileTime) GetProcAddress(hKernel32, "GetSystemTimePreciseAsFileTime");
-#if defined(__GNUC__)
+#if defined(__GNUC__) && (__GNUC__ == 8)
 #pragma GCC diagnostic pop
 #endif
 

